### PR TITLE
Add to Codegen Interface, Refactor Proto Serialization

### PR DESCRIPTION
**Pull request by ncteisen**
A bit early on this, but I want to sanity check these changes.

### CODE REVIEW

**Comment by ncteisen** [comment]
Would be nice to have this in the next import. Though it could wait till the next next if need be. The only constraint is that this be in for the first import AFTER the internal Cord optimization ~lands~ is merged.

**Comment by grpc-testing** [comment]
```
No significant performance differences
```

File: include/grpc++/impl/codegen/proto_utils.h [hunk 1]
```diff
@@ -175,64 +175,87 @@ class GrpcBufferReader final
...
- private:
+ protected:
   int64_t byte_count_;
   int64_t backup_count_;
   grpc_byte_buffer_reader reader_;
   grpc_slice slice_;
   Status status_;
 };
+
+// BufferWriter must be a subclass of io::ZeroCopyOutputStream.
```
Comment (ctiller) [inline include/grpc++/impl/codegen/proto_utils.h]: static_assert(std::is_base_of(BufferWriter, io::ZeroCopyOutputStream), "BufferWriter must be a subclass of io::ZeroCopyOutputStream");

File: include/grpc++/impl/codegen/proto_utils.h [hunk 2]
```diff
@@ -175,64 +175,87 @@ class GrpcBufferReader final
...
+  } else {
+    BufferWriter writer(bp, internal::kGrpcBufferWriterMaxBufferLength);
+    return msg.SerializeToZeroCopyStream(&writer)
+               ? g_core_codegen_interface->ok()
+               : Status(StatusCode::INTERNAL, "Failed to serialize message");
+  }
+}
+
+// BufferReader must be a subclass of io::ZeroCopyInputStream.
+template <class BufferReader, class T>
```
Comment (ctiller) [inline include/grpc++/impl/codegen/proto_utils.h]: Same static assert here

**Pull request merged**